packages/snaps-controllers/src/snaps/SnapController.ts

```diff
@@ -2727,37 +2734,37 @@ export class SnapController extends BaseController<
       "'permissionName' must be either a string or null.",
     );
 
-    if (permissionName) {
-      const hasPermission = this.messagingSystem.call(
-        'PermissionController:hasPermission',
-        snapId,
-        permissionName,
-      );
+    const permissions = this.messagingSystem.call(
+      'PermissionController:getPermissions',
+      snapId,
+    );
 
-      if (!hasPermission) {
-        throw new Error(
-          `Snap "${snapId}" is not permitted to use "${permissionName}".`,
-        );
-      }
+    // If permissionName is null, the handler does not require a permission.
+    if (
+      permissionName !== null &&
+      (!permissions || !hasProperty(permissions, permissionName))
+    ) {
+      throw new Error(
+        `Snap "${snapId}" is not permitted to use "${permissionName}".`,
+      );
     }
 
+    const handlerPermissions = permissionName
+      ? // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
```

Comment:
Can `// eslint-disable-next-line @typescript-eslint/no-non-null-assertion` be avoided? With something like:
```typescript
const handlerPermissions = permissionName
      ? permissions?.[permissionName]
      : undefined;
```

Reply:
If `permissionName` is not null and `permissions` is falsy the above if-statement should throw. 

I don't think it makes sense to add `?.` if we are sure it will never be null. WDYT?

Reply:
Alright, I agree that it would not make much more sense with adding `?.` given that the `if` statement is checking the `permissions`. But, still, then the override with `!` and `eslint-disable-next-line` gets in the way and looks like hacking.

So, if we are sure that **it will never be null** and we know **what** it is, then we can assert its type and get rid of the `eslint-disable`:
```typescript
const handlerPermissions = permissionName
      ? (permissions as SubjectPermissions<PermissionConstraint>)[
          permissionName
        ]
      : undefined;
```

Reply:
Sounds good to me, pushed!